### ISSUE TITLE
Cannot find THCTensor.hpp (required indirectly by THCTensorMathReduce.hpp) when building extension

### ISSUE THREAD

**Issue by mcarilli** [pytorch/pytorch#8267]
## Issue description

When building my extensions (and preparing to upstream eventually) I'd like to be able to use the functors present in THCTensorMathReduce.cuh.  Until recently (last week ish) it worked fine, but now #include THC/THCTensorMathReduce.cuh fails.  The include chain is
THCTensorMathReduce.cuh -> THCReduce.cuh -> THCTensorTypeUtils.cuh -> THCTensor.hpp -> failure.

I installed the latest Pytorch master via "python setup.py install."  The resulting include directory on my system (in my case /home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/THC) contains THCTensorMathReduce.cuh, THCReduce.cuh, and THCTensorTypeUtils.cuh, but not THCTensor.hpp.

I believe the offending commit, after which compilation fails, is https://github.com/pytorch/pytorch/commit/61fd99e1b313c0bdb607f31a443d2ce027aed582#diff-596ad5d24f67dca5412b805737911064.

It may be worth adding a test for including this file (and any other headers we anticipate being useful to extension builders) in the cpp extension CI.  If you think so, I'm happy to flesh out my sample for the purpose.

@ngimel @csarofeen @mruberry for camaraderie, @goldsborough because he seems to be the extension guy, @gchanan because he's the author of the commit I suspect is at fault, although I could be wrong.

## Code example

I attached a minimal working repro:  [tensormathreduce.tar.gz](https://github.com/pytorch/pytorch/files/2082590/tensormathreduce.tar.gz).  On a system with nvcc and the latest Pytorch master, try
```
tar -zxvf tensormathreduce.tar.gz
cd tensormathreduce
python setup.py install  
```
You should see something along the lines of 
/usr/local/cuda/bin/nvcc -I/home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include -I/home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/TH -I/home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/THC -I/usr/local/cuda/include -I/home/mcarilli/anaconda3/include/python3.6m -c backend.cu -o build/temp.linux-x86_64-3.6/backend.o -O2 -DTORCH_EXTENSION_NAME=include_test_C --compiler-options '-fPIC' -std=c++11
In file included from /home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/THC/THCReduce.cuh:11:0,
                 from /home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/THC/THCTensorMathReduce.cuh:7,
                 from backend.cu:5:
/home/mcarilli/anaconda3/lib/python3.6/site-packages/torch/lib/include/THC/THCTensorTypeUtils.cuh:8:25: fatal error: THCTensor.hpp: No such file or directory
compilation terminated.

If you comment out the line "#include <THC/THCTensorMathReduce>" in backend.cu, compilation should succeed.

## System Info
Collecting environment information...
PyTorch version: 0.5.0a0+0c9b5f0
Is debug build: No
CUDA used to build PyTorch: 9.1.85

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.9.4

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: 
GPU 0: TITAN X (Pascal)
GPU 1: Graphics Device

Nvidia driver version: 390.48
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5.real
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudnn.so
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudnn.so.7.1.4
/usr/local/cuda-9.1/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip3] numpy (1.13.3)
[pip3] numpydoc (0.7.0)
[pip3] torch (0.5.0a0+0c9b5f0)
[pip3] torch-test-cpp-extension (0.0.0)
[pip3] torchvision (0.2.0)
[conda] magma-cuda90              2.2.0                hae16b58_1    pytorch
[conda] torch                     0.5.0a0+0c9b5f0           <pip>
[conda] torch-test-cpp-extension  0.0.0                     <pip>
[conda] torchvision               0.2.0                     <pip>

**Comment by goldsborough** [comment]
Hi, including anything TH in C++ extensions is unsupported. TH is a very low level backend in Pytorch and anything can be deleted or broken any day. It's not part of ATen's API and therefore also not part of the C++ extension API. You can find a way to hack it yourself but otherwise this is a won'tfix from us. Sorry!

**Comment by soumith** [comment]
@mcarilli is from nvidia, they've been working on our codebase for a long time.

@gchanan can you tell us what's the path forward for them to be productive. I know you are planning to slowly nuke THC.
- Will DeviceTensor and TypeUtils etc. be available via ATen, or are they simply going away?
- What are the steps you are taking to refactor out THC, what should @mcarilli @ngimel @mruberry and @csarofeen be aware of?
- @mcarilli is working on revamping the Embedding backward kernels to make them much faster. Anything he should be aware of for this short-term work?

**Comment by ngimel** [comment]
Then why is THCTensorMathReduce.cuh is installed in site-packages/torch where it looks like it is includable?

**Comment by mcarilli** [comment]
@goldsborough In this case I was using THCTensorMathReduce functors to be as "pytorchic" as possible.  If you don't advise that, I can easily roll my own.  That being said, you can't really write custom Cuda extensions without dealing with raw data pointers, for which THC utilities/type conversions come in handy, so the strategy wrt to THC is definitely something we're interested to hear.

@soumith The kernel I'm writing for embedding backward will be a drop-in replacement for embedding_backward_feature_kernel in Embedding.cu, with no changes required by the C++ wrapper signatures, so the build system should be perfectly happy with it.  I feel comfortable enough with the ATen API and dispatch to adjust wrapper-side C++ without bothering your guys in the immediate term, unless there's something important you feel I should know.  Of course I appreciate the offer of help and I'll ask if I hit any roadblocks.

@goldsborough The deprecated shuffles will be removed as part of this change, and I expect a similar solution will work to remove the deprecated shuffles in THCUNN/LookupTable.cu, which I'll tackle next.

**Comment by gchanan** [comment]
Some good questions here.

> Then why is THCTensorMathReduce.cuh is installed in site-packages/torch where it looks like it is includable?

I don't think I fully understand how this works.  @goldsborough does C++ extensions install different headers than pytorch or is THTensor.hpp not being installed with THCTensorMathReduce.cuh for some other reason (maybe the cmake is wrong)?  While I agree with @goldsborough that things at the TH/THC level are internal APIs, it doesn't make sense to me why one of these files is included and not the other?

> Will DeviceTensor and TypeUtils etc. be available via ATen, or are they simply going away?

TypeUtils is definitely going away.  It can't work going forward (it is templatized on the Tensor type for which there will be only one) and there will be better APIs that replace it.  For DeviceTensor I haven't decided yet -- if someone could provide a writeup of what they find particularly useful about it, I can see about porting it to ATen.

> What are the steps you are taking to refactor out THC, what should @mcarilli @ngimel @mruberry and @csarofeen be aware of?

The problem is we are in an in-between state.  the ATen API is public but limited; it doesn't have all the utility functions of TH/THC and you don't have direct control over the sizes/strides/dimensions, so you actually _can't_ do some things without TH/THC.  That's what the work @ezyang and I are doing is attempting to address: ATen and TH/THC will be the same thing in that the API will look like ATen but you'll have direct control over sizes/strides/dimensions, and better public APIs.  But we aren't there yet.

The major steps are roughly:

Step 1.  There will only be a single THTensor and a single THCTensor type.  So that means things that templatize on the type of THTensor/THCTensor will not work.

Step 2.  These (THTensor/THCTensor) types will eventually change to be the C10/A10 equivalent.  Basically, everything will look and behave like ATen, except you'll also have direct control over sizes/strides/dimensions that we don't provide at the ATen level today.

Step 3.  Move any APIs that are publicly useful and stable in TH/THC out to ATen.  This is much easier once Step 1 and 2 are done because ideally it's just a namespace change, because the Tensor types/APIs are identical.

> @mcarilli is working on revamping the Embedding backward kernels to make them much faster. Anything he should be aware of for this short-term work?

Honestly I'd just discuss with us either here on or slack what you need.  As I said before, it may be possible to port DeviceTensor to ATen.  Alternatively, I don't think it's the worst thing (until we finish the work above), to use a limited set of TH/THC APIs with the understanding that they _will_ break.  But maybe @goldsborough has a better idea.

**Comment by goldsborough** [comment]
FWIW @fmassa has been using TH things in his C++ extensions, which we agreed upon since he is a longstanding TH expert and very comfortable with TH, under the condition that this is at his own risk given our ongoing revamp of our internals. I've been updated that @mcarilli and company are similarly TH veterans, so I would be fine im opening an unofficial niche for you to access TH in extensions. 

To clarify: Officially, the public extension API is exclusively what is included in torch/torch.h, which is primarily ATen, pybind11 and some autograd specific headers (such as pybind bindings for Variables). No other headers are documented as being part of the API and their existence in the installed package doesn't influence this. I think we copied them into the package at the time so that @fmassa could use them unofficially. If there is a header missing we can install it for you, but again this would be an unofficial agreement between us. If we advertised TH as part of the extension API, folks like @gchanan would lose their freedom to improve ATen, especially given our ongoing caffe2/c10 consolidation efforts.

I think the best line of action is:
1. Can you easily implement what you want only using ATen?
2. If not, can we easily extend ATen for you to maybe add a particular function you need?
3. If not, we can add missing TH headers and you are free to use TH in extensions since you are special contributors, but unofficially and at the risk that we will be forced to remove certain headers in the future.

I'm fine with any of the above and I'll be happy to help with whatever unblocks you soonest.

Does this sound reasonable?

**Comment by csarofeen** [comment]
@goldsborough I think options 1 is fine in theory unless we run into something that we need from TH that wasn't supported in at the 0.4 ATen cut. I'm fine with 2, if we can get what we have working with both 0.4 and master in a reasonable manner. If not I may come back and ask for 3. Curious if @fmassa has any hard breaks since 0.4 in his extension.

**Comment by mcarilli** [comment]
@goldsborough @soumith I've gotten our stuff working with what's currently available, and now we know what to watch out for, so you can close this issue if you think that's appropriate.

**Comment by andrei-pokrovsky** [comment]
Any thoughts on ABI issues with C++ binary API?
